### PR TITLE
Solving issues

### PR DESCRIPTION
**Pull request by PovilasDuo**
1. Some exceptions are handled, but still throwable actions may happen unexpectedly. For example, str.split() can throw exceptions, that is not explicitly handled.

removed validateInput and other methods were reworked to solve this

2. The checkIfFilePathIsRelative() and makePathsAbsolute() logic is a bit redundant. Libs usually resolve it automatically.

Replaced with one-liner from os library - os.Path(filePath, os.pwd)

3. You use both os.Path and java.nio.file.Paths, would be best to use the same type everywhere instead of mixing different libs for the same purpose.

I am not sure about this - I've checked my project and I seems that I only use 'java.nio.file.NoSuchFileException'

Could you elaborate on this one?

4. Function validateOutput() not only returns Unit (instead of boolean for example), but it also doesn't even throw error, just prints. So essentially, it doesn't really validate anything.

Removed the logic since the core purpose - to notify if there were no matches is implied with an empty output file

5. While on the topic of printing - leaving println() statements is usually not recommended in production code. If you really need to print something, using a proper logger is recommended.

Used proper logging with Logger library

6. The JSON parsers are quite 'low-level' or at least a bit overcomplicated. We are not very familiar with this library, so it might just be the fault of the lib.

uPickle is more simple if there are no custom data Types. With results it is enough to add 'derives ReadWriter' (Results class). This seems, for me atleast, to be a normal parser

7.Tests create files in the root dir of the project, and those files stay there after the tests are ran. If you do need to write to the filesystem, write to some temporary directory instead of to the repo.

Added a method which removes them after the tests

8. windingNumberPointPolygon(polygon, loc.coordinates) != 0 that's scattering algorithmic logic around the code. We'd expect a method clearly containing everything about matching a point to a polygon.

windingNumberPointPolygon(polygon, loc.coordinates) is now of type Boolean - returns true if the point exists in a polygon. Also renamed the method to better depict it's purpose

9.You could use an argument parsing library instead of writing your own.

Implemented argument parsing via scopt

10. We would recommend using one standard way of passing inputs into your unit tests. One of them could be using a special resources folder https://www.baeldung.com/scala/read-file-from-resources.

Implemented resource input reading

11. When intercepting an exception do you need to assign it to a val, if you are not using it later?

Removed unnecessary assignments

12. Do you think you could extend the test coverage (testing more edge cases, improper data...).

Yes

⡴⠑⡄⠀⠀⠀⠀⠀⠀⠀ ⣀⣀⣤⣤⣤⣀⡀
⠸⡇⠀⠿⡀⠀⠀⠀⣀⡴⢿⣿⣿⣿⣿⣿⣿⣿⣷⣦⡀
⠀⠀⠀⠀⠑⢄⣠⠾⠁⣀⣄⡈⠙⣿⣿⣿⣿⣿⣿⣿⣿⣆
⠀⠀⠀⠀⢀⡀⠁⠀⠀⠈⠙⠛⠂⠈⣿⣿⣿⣿⣿⠿⡿⢿⣆
⠀⠀⠀⢀⡾⣁⣀⠀⠴⠂⠙⣗⡀⠀⢻⣿⣿⠭⢤⣴⣦⣤⣹⠀⠀⠀⢀⢴⣶⣆
⠀⠀⢀⣾⣿⣿⣿⣷⣮⣽⣾⣿⣥⣴⣿⣿⡿⢂⠔⢚⡿⢿⣿⣦⣴⣾⠸⣼⡿
⠀⢀⡞⠁⠙⠻⠿⠟⠉⠀⠛⢹⣿⣿⣿⣿⣿⣌⢤⣼⣿⣾⣿⡟⠉
⠀⣾⣷⣶⠇⠀⠀⣤⣄⣀⡀⠈⠻⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡇
⠀⠉⠈⠉⠀⠀⢦⡈⢻⣿⣿⣿⣶⣶⣶⣶⣤⣽⡹⣿⣿⣿⣿⡇
⠀⠀⠀⠀⠀⠀⠀⠉⠲⣽⡻⢿⣿⣿⣿⣿⣿⣿⣷⣜⣿⣿⣿⡇
⠀⠀ ⠀⠀⠀⠀⠀⢸⣿⣿⣷⣶⣮⣭⣽⣿⣿⣿⣿⣿⣿⣿⠇
⠀⠀⠀⠀⠀⠀⣀⣀⣈⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠇
⠀⠀⠀⠀⠀⠀⢿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠃

### CODE REVIEW

**Comment by arnasbr** [comment]
> 3. You use both os.Path and java.nio.file.Paths, would be best to use the same type everywhere instead of mixing different libs for the same purpose.
>
> I am not sure about this - I've checked my project and I seems that I only use 'java.nio.file.NoSuchFileException'

Seems like you're correct, but still - I believe you could just use `java.nio.file` for everything, including reading, etc. But it's not a huge deal

File: src/main/scala/models/Location.scala [hunk 1]
```diff
@@ -1,37 +1,39 @@
...
 }
 
 object Location {
-  private def fromList(name: String, coordinatesList: List[Double]) : Option[Location] = coordinatesList match {
-    case x::y::Nil => Some(Location(name, Point(x,y)))
-    case _ => None
+  private def fromList(name: String, coordinatesList: List[Double]) : ValidatedNel[String, Location] =
+    coordinatesList match {
+      case x::y::Nil => Validated.validNel(Location(name, Point(x,y)))
+      case _ => Validated.invalidNel("Expected exactly two coordinates (x, y)")
```
Comment (arnasbr) [inline src/main/scala/models/Location.scala:16]: Nice application of pattern matching

File: src/test/scala/RegionTest.scala [hunk 2]
```diff
@@ -22,10 +23,12 @@ class RegionTest extends AnyFunSuite{
       "Not all regions were added when reading the file:" +
         "\nExpected: regions list length: 3" +
         s"\nActual: regions list length: ${regions.length}")
+
+    cleanUpFiles(List(path))
```
Comment (arnasbr) [inline src/test/scala/RegionTest.scala:27]: Do you think you actually need to write to files here? Would it be possible to parse the json directly from the input string? If so, how would your code have to change?

**Review by arnasbr — APPROVED** [review]
Nice improvements generally :) we will discuss your solution in more detail live

**Comment by gergelytraveltime** [comment]
> > Seems like you're correct, but still - I believe you could just use `java.nio.file` for everything, including reading, etc. But it's not a huge deal
> 
> Are there any benefits of switching from 'os-lib' to 'java.nio.file'?

In a certain sense I suppose `os-lib` is a higher-level lib that java.nio.file. I don't have any hard preference. Could you implement file actions with one of them? (Is it possible with `os-lib`?)

File: src/test/scala/LocationTest.scala [hunk 3]
```diff
@@ -42,5 +43,37 @@ class LocationTest extends AnyFunSuite {
...
+  }
+
+  test("Read Locations - string coordinates") {
+    val badJson = """[{ "name": "BadCoordStr", "coordinates": ["a", "b"] }]"""
+    intercept[Exception](readLocations(badJson))
+  }
+
+  test("Read Locations - empty string name") {
+    val badJson = """[{ "name": "", "coordinates": [-0.1276, 51.5072] }]"""
+    intercept[IllegalArgumentException](readLocations(badJson))
```
Comment (gergelytraveltime) [inline src/test/scala/LocationTest.scala:77]: This one is enough specific already :)
